tests/conformance/test_conformance.py

```diff
@@ -0,0 +1,449 @@
+
+
+def get_iam_permissions(client, _preconditions, bucket):
+    bucket = client.bucket(bucket.name)
+    permissions = ["storage.buckets.get", "storage.buckets.create"]
+    bucket.test_iam_permissions(permissions)
+
+
+def get_service_account_email(client, _preconditions):
+    client.get_service_account_email()
+
+
+def make_bucket_public(client, _preconditions, bucket):
+    bucket = client.bucket(bucket.name)
+    bucket.make_public()
+
+
+def delete_blob(client, _preconditions, bucket, object):
+    bucket = client.bucket(bucket.name)
+    if _preconditions:
+        generation = object.generation
+        bucket.delete_blob(object.name, if_generation_match=generation)
+    else:
+        bucket.delete_blob(object.name)
+
+
+def lock_retention_policy(client, _preconditions, bucket):
+    bucket2 = client.bucket(bucket.name)
+    bucket2.retention_period = 60
+    bucket2.patch()
+    bucket2.lock_retention_policy()
+
+
+def patch_bucket(client, _preconditions, bucket):
+    bucket = client.get_bucket("bucket")
+    metageneration = bucket.metageneration
+    bucket.storage_class = "COLDLINE"
+    if _preconditions:
+        bucket.patch(if_metageneration_match=metageneration)
+    else:
+        bucket.patch()
+
+
+def update_bucket(client, _preconditions, bucket):
+    bucket = client.get_bucket("bucket")
+    metageneration = bucket.metageneration
+    bucket._properties = {"storageClass": "STANDARD"}
+    if _preconditions:
+        bucket.update(if_metageneration_match=metageneration)
+    else:
+        bucket.update()
+
+
+def patch_blob(client, _preconditions, bucket, object):
+    blob = client.bucket(bucket.name).blob(object.name)
+    blob.metadata = {"foo": "bar"}
+    if _preconditions:
+        blob.patch(if_metageneration_match=object.metageneration)
+    else:
+        blob.patch()
+
+
+def update_blob(client, _preconditions, bucket, object):
+    blob = client.bucket(bucket.name).blob(object.name)
+    blob.metadata = {"foo": "bar"}
+    if _preconditions:
+        blob.update(if_metageneration_match=object.metageneration)
+    else:
+        blob.update()
+
+
+def copy_blob(client, _preconditions, bucket, object):
+    bucket = client.bucket(bucket.name)
+    destination = client.bucket("bucket")
+    if _preconditions:
+        bucket.copy_blob(
+            object, destination, new_name=uuid.uuid4().hex, if_generation_match=0
+        )
+    else:
+        bucket.copy_blob(object, destination)
+
+
+def rename_blob(client, _preconditions, bucket, object):
+    bucket = client.bucket(bucket.name)
+    new_name = uuid.uuid4().hex
+    if _preconditions:
+        bucket.rename_blob(object, new_name, if_generation_match=0)
+    else:
+        bucket.rename_blob(object, new_name)
+
+
+def rewrite_blob(client, _preconditions, bucket, object):
+    new_blob = client.bucket(bucket.name).blob(uuid.uuid4().hex)
+    new_blob.metadata = {"foo": "bar"}
+    if _preconditions:
+        new_blob.rewrite(object, if_generation_match=0)
```

Comment:
Should this use generation = 0 or the generation from the existing object?

Reply:
`if_generation_match=0` works when there are no live versions of the destination object. The destination blob `new_blob` is just instantiated at this point, so the operation will succeed.